core/common/src/main/java/alluxio/underfs/NonAtomicFileOutputStream.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.underfs.options.NonAtomicCreateOptions;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A {@link NonAtomicFileOutputStream} writes to a temporary file and renames on close.
+ */
+@NotThreadSafe
+public class NonAtomicFileOutputStream extends OutputStream {
```

Comment:
At the moment, FileOutStream (the Alluxio client facing API) is different from the UFS OutputStream API. NonAtomicFileOutputStream is the non atomic version of the UFS OutputStream API. 

I agree that the two should ideally converge to better support the write through case. To be more specific, the difference is getBytesWritten and cancel. I see the benefit of cancel, as @calvinjia pointed out. The benefit of a UFS getBytesWritten, I don't know.